### ISSUE TITLE
Allow --no-single-branch git option

### ISSUE THREAD

**Issue by magaliruffier** [travis-ci/travis-ci#4806]
Since git release 1.8.3.2, using --depth ignores remote branches unless --no-single-branch is also specified
As there is no way to remove the --depth option in the git set up, would it be possible to allow the --no-single-branch option to allow the same behaviour as before 1.8.3.2?

An example use case:
I want to compare files between branches, the master branch which is being tested, and the branch from a previous release
I used to do this with:

On a side note, I could not find any documentation that lists what options can be specified for git
I know I can use
git:
 depth: 50
but am stuck beyond that

**Comment by vladbarosan** [comment]
Is there any update on this issue?

**Comment by BanzaiMan** [comment]
Does running the following command achieve what you need?

    git fetch —unshallow

**Comment by skarip** [comment]
Not alone but combined with updating the config will undo the single branch AND shallow clone:
```
git config remote.origin.fetch refs/heads/*:refs/remotes/origin/*
git fetch --unshallow
```

This is still very much not ideal compared to just doing a shallow clone but with the --no-single-branch flag because the above fetch --unshallow will bring with it all the commit history (in my case the difference being a ~50% size increase over the clone with depth and no single branch)

**Comment by BanzaiMan** [comment]
You can remove `--depth` flag with `git.depth: false`. See https://github.com/travis-ci/travis-build/pull/1093.

**Comment by BanzaiMan** [comment]
https://docs.travis-ci.com/user/customizing-the-build#Git-Clone-Depth

**Comment by skarip** [comment]
Yup, that will definitely solve the proximate problem, but there's a performance issue with an unshallow clone of the repository. It's important to differentiate a shallow clone and a single branch clone. The former will copy all commits in the history while the latter restricts the fetch refs (in the .git/config) to a single branch. A shallow clone without the single branch flag (or with the no-single-branch flag) will be a much faster and lighter clone than a clone with no depth (unshallow).

**Comment by drd** [comment]
I would also like to have the `--no-single-branch` as an option, since without it I will need to change how I manage release builds to generate proper release notes.
